src/Swarm/Game/State.hs

```diff
@@ -362,19 +362,23 @@ clearFocusedRobotLogUpdated = do
   n <- use focusedRobotID
   robotMap . ix n . robotLogUpdated .= False
 
--- | Add a robot to the game state, generating a unique ID number for
---   it, also adding it to the index of robots by location, and return
---   the robot updated with its unqiue ID.
-addRobot :: Has (State GameState) sig m => Robot -> m Robot
+-- | Add a robot to the game state, adding it to the main robot map,
+--   the active robot set, and to to the index of robots by
+--   location. If it doesn't already have a unique ID number, generate
+--   one for it.
+addRobot :: Has (State GameState) sig m => Robot -> m ()
 addRobot r = do
-  rid <- gensym <+= 1
+  r' <- case r ^. robotID of
+    (-1) -> do
```

Comment:
Hmm, you're right, it is hacky.  But I am reluctant to change the robot ID to `Maybe Nat` because it would mean having to annoyingly deal with the `Nothing` possibility in many, many places in the code where it is not supposed to be possible.

Hmm, maybe it would actually be worthwhile doing something like what you suggested before, where the `Robot` type is somehow indexed to show whether it has a robot ID or not (i.e. keeping track of the "robot lifecycle"). Something like this, perhaps?
```
data RobotRecord f = RobotRecord { ..., _robotID :: f Int, ... }
type Robot = RobotRecord Identity
type URobot = RobotRecord Maybe  --- U for Unidentified
mkRobot :: Int -> ... -> Robot
mkURobot :: ... -> URobot
```
I'd have to think more carefully about where each of these types would go, and how they would interact with the places we are currently using a robot ID of (-1).

Reply:
So I started doing this refactoring, changing the `Robot` type to track whether or not it had an ID number at the type level, then just fixing type errors one by one to complete the refactoring.  Most of the errors were of the usual "refactoring" kind: changing all the places I was referring to a name that is now called something different, etc.  But then at one point it threw up a type mismatch error, saying that it was expecting a certain list to have the type `[Robot]` (i.e. robots with IDs) but it actually had the type `[URobot]` (robots without IDs)... and it was actually a bug!  We weren't assigning ID numbers to robots read from challenge descriptions, so in challenges with multiple robots all but one would instantly disappear (since they all shared the ID number of 1).  It is now impossible to add a robot without an ID number to the world, because the types simply don't match.  (Of course it is still theoretically possible to add multiple robots with the same ID, but in practice you have to add robots via the `addURobot` function which will automatically assign a fresh ID).

A big thank you to @polux for bringing me to my senses.  I will open a PR with the changes shortly.